// @zeitgeistpm/indexer/dist/index.mjs
import { GraphQLClient } from 'graphql-request';
import { augments } from './augments/index.mjs';
import { getSdk } from './graphql/sdk.mjs';
export { AccountBalanceOrderByInput, AccountBalancesDocument, AccountOrderByInput, AssetOrderByInput, AssetsDocument, FullAccountBalanceFragmentDoc, FullAssetFragmentDoc, FullHistoricalAccountBalanceFragmentDoc, FullHistoricalAssetsFragmentDoc, FullMarketFragmentDoc, FullPoolFragmentDoc, HistoricalAccountBalanceOrderByInput, HistoricalAccountBalancesDocument, HistoricalAssetOrderByInput, HistoricalAssetsDocument, HistoricalMarketOrderByInput, HistoricalPoolOrderByInput, MarketOrderByInput, MarketStatusCountDocument, MarketsDocument, PingQueryDocument, PoolOrderByInput, PoolsDocument, getSdk } from './graphql/sdk.mjs';

/**
 * Create a zeitgeist indexer client.
 *
 * @param config Config
 * @returns ZeitgeistIndexer
 */
const create = (config) => {
    const client = new GraphQLClient(config.uri, { fetch: globalThis.fetch });
    const gql = getSdk(client);
    return {
        client,
        ...gql,
        ...augments(gql),
    };
};

export { create };
