// @zeitgeistpm/indexer/dist/index.js
'use strict';

var graphqlRequest = require('graphql-request');
var index = require('./augments/index.js');
var sdk = require('./graphql/sdk.js');

/**
 * Create a zeitgeist indexer client.
 *
 * @param config Config
 * @returns ZeitgeistIndexer
 */
const create = (config) => {
    const client = new graphqlRequest.GraphQLClient(config.uri, { fetch: globalThis.fetch });
    const gql = sdk.getSdk(client);
    return {
        client,
        ...gql,
        ...index.augments(gql),
    };
};

Object.defineProperty(exports, 'AccountBalanceOrderByInput', {
    enumerable: true,
    get: function () { return sdk.AccountBalanceOrderByInput; }
});
exports.AccountBalancesDocument = sdk.AccountBalancesDocument;
Object.defineProperty(exports, 'AccountOrderByInput', {
    enumerable: true,
    get: function () { return sdk.AccountOrderByInput; }
});
Object.defineProperty(exports, 'AssetOrderByInput', {
    enumerable: true,
    get: function () { return sdk.AssetOrderByInput; }
});
exports.AssetsDocument = sdk.AssetsDocument;
exports.FullAccountBalanceFragmentDoc = sdk.FullAccountBalanceFragmentDoc;
exports.FullAssetFragmentDoc = sdk.FullAssetFragmentDoc;
exports.FullHistoricalAccountBalanceFragmentDoc = sdk.FullHistoricalAccountBalanceFragmentDoc;
exports.FullHistoricalAssetsFragmentDoc = sdk.FullHistoricalAssetsFragmentDoc;
exports.FullMarketFragmentDoc = sdk.FullMarketFragmentDoc;
exports.FullPoolFragmentDoc = sdk.FullPoolFragmentDoc;
Object.defineProperty(exports, 'HistoricalAccountBalanceOrderByInput', {
    enumerable: true,
    get: function () { return sdk.HistoricalAccountBalanceOrderByInput; }
});
exports.HistoricalAccountBalancesDocument = sdk.HistoricalAccountBalancesDocument;
Object.defineProperty(exports, 'HistoricalAssetOrderByInput', {
    enumerable: true,
    get: function () { return sdk.HistoricalAssetOrderByInput; }
});
exports.HistoricalAssetsDocument = sdk.HistoricalAssetsDocument;
Object.defineProperty(exports, 'HistoricalMarketOrderByInput', {
    enumerable: true,
    get: function () { return sdk.HistoricalMarketOrderByInput; }
});
Object.defineProperty(exports, 'HistoricalPoolOrderByInput', {
    enumerable: true,
    get: function () { return sdk.HistoricalPoolOrderByInput; }
});
Object.defineProperty(exports, 'MarketOrderByInput', {
    enumerable: true,
    get: function () { return sdk.MarketOrderByInput; }
});
exports.MarketStatusCountDocument = sdk.MarketStatusCountDocument;
exports.MarketsDocument = sdk.MarketsDocument;
exports.PingQueryDocument = sdk.PingQueryDocument;
Object.defineProperty(exports, 'PoolOrderByInput', {
    enumerable: true,
    get: function () { return sdk.PoolOrderByInput; }
});
exports.PoolsDocument = sdk.PoolsDocument;
exports.getSdk = sdk.getSdk;
exports.create = create;
