// @zeitgeistpm/sdk/dist/meta/market/types.js
'use strict';

var superstruct = require('superstruct');

/**
 * IO validation of Market Categories using superstruct
 */
const IOMarketCategoryMetadata = superstruct.object({
    name: superstruct.string(),
    ticker: superstruct.optional(superstruct.string()),
    img: superstruct.optional(superstruct.string()),
    color: superstruct.optional(superstruct.string()),
});
/**
 * IO Scalar range type
 */
superstruct.union([superstruct.literal('date'), superstruct.literal('number')]);
/**
 * IO validation of Market metadata object using superstruct
 */
const IOMarketMetadata = superstruct.object({
    __meta: superstruct.literal('markets'),
    slug: superstruct.string(),
    description: superstruct.string(),
    question: superstruct.string(),
    tags: superstruct.optional(superstruct.array(superstruct.string())),
    confidentialId: superstruct.optional(superstruct.string()),
    img: superstruct.optional(superstruct.string()),
    scalarType: superstruct.optional(superstruct.union([superstruct.literal('date'), superstruct.literal('number')])),
    categories: superstruct.optional(superstruct.array(IOMarketCategoryMetadata)),
});

exports.IOMarketCategoryMetadata = IOMarketCategoryMetadata;
exports.IOMarketMetadata = IOMarketMetadata;
