// @zeitgeistpm/utility/dist/time/timespan.esm.js
import { isNumber } from '@polkadot/util';
import { dateBlock, blockDate } from './time.esm.js';

/**
 * Typeguard for BlockTimespan
 *
 * @param timespan Timespan
 * @returns timespan is BlockTimespan
 */
const isBlocks = (timespan) => isNumber(timespan.start) && isNumber(timespan.end);
/**
 * Typeguard for DateTimespan
 *
 * @param timespan Timespan
 * @returns timespan is DateTimespan
 */
const isDates = (timespan) => timespan.start instanceof Date && timespan.end instanceof Date;
function from(now, timespan) {
    return isDates(timespan) ? datesToBlocks(now, timespan) : blocksToDates(now, timespan);
}
/**
 * Convert timespan to blocks.
 *
 * @param ctx RpcContext
 * @param timespan Timespan
 * @returns Promise<BlockTimespan>
 */
const asBlocks = (now, timespan) => isDates(timespan) ? from(now, timespan) : timespan;
/**
 * Convert a timespan of dates to block range.
 *
 * @note Is approximate based on chain block production time and not exact.
 *
 * @param ctx RpcContext
 * @param dates DateTimespan
 * @returns Promise<BlockTimespan>
 */
const datesToBlocks = (now, dates) => {
    return {
        start: dateBlock(now, dates.start),
        end: dateBlock(now, dates.end),
    };
};
/**
 * Convert a timespan of dates to block range.
 *
 * @note Is approximate based on chain block production time and not exact.
 *
 * @param ctx RpcContext
 * @param dates DateTimespan
 * @returns Promise<BlockTimespan>
 */
const blocksToDates = (now, dates) => {
    return {
        start: blockDate(now, dates.start),
        end: blockDate(now, dates.end),
    };
};
/**
 * Get the duration of a timespan in milliseconds.
 *
 * @param time ChainTime
 * @param timeSpan Timespan
 * @returns number
 */
const toMs = (time, timeSpan) => {
    return isDates(timeSpan)
        ? timeSpan.end.getTime() - timeSpan.start.getTime()
        : (timeSpan.end - timeSpan.start) * time.period;
};

export { asBlocks, blocksToDates, datesToBlocks, from, isBlocks, isDates, toMs };
