// @zeitgeistpm/sdk/dist/index.js
'use strict';

var types = require('./types.js');
var index = require('./config/known/index.js');
var types$1 = require('./config/types.js');
var types$6 = require('./meta/types.js');
var types$7 = require('./meta/market/types.js');
var create = require('./create.js');
var ztg = require('./primitives/ztg.js');
var assetid = require('./primitives/assetid.js');
var blocknumber = require('./primitives/blocknumber.js');
var marketid = require('./primitives/marketid.js');
var data = require('./primitives/data.js');
var na = require('./primitives/na.js');
var storage = require('./config/known/storage.js');
var types$2 = require('./config/known/types.js');
var types$3 = require('./context/types.js');
var types$4 = require('./model/markets/functions/create/types.js');
var market = require('./model/markets/market.js');
var marketstage = require('./model/markets/marketstage.js');
var types$5 = require('./model/swaps/functions/getPool/types.js');
var pool = require('./model/swaps/pool.js');



exports.isFullSdk = types.isFullSdk;
exports.isIndexedSdk = types.isIndexedSdk;
exports.isRpcSdk = types.isRpcSdk;
exports.sdk = types.sdk;
exports.batterystation = index.batterystation;
exports.batterystationIndexer = index.batterystationIndexer;
exports.batterystationRpc = index.batterystationRpc;
exports.indexers = index.indexers;
exports.mainnet = index.mainnet;
exports.mainnetIndexer = index.mainnetIndexer;
exports.mainnetRpc = index.mainnetRpc;
exports.rpcs = index.rpcs;
exports.asIndexerConfig = types$1.asIndexerConfig;
exports.asRpcConfig = types$1.asRpcConfig;
exports.isFullConfig = types$1.isFullConfig;
exports.isIndexerConfig = types$1.isIndexerConfig;
exports.isRpcConfig = types$1.isRpcConfig;
exports.createStorage = types$6.createStorage;
exports.saturate = types$6.saturate;
exports.tagged = types$6.tagged;
exports.IOMarketCategoryMetadata = types$7.IOMarketCategoryMetadata;
exports.IOMarketMetadata = types$7.IOMarketMetadata;
exports.create = create.create;
exports.create$ = create.create$;
exports.createFullContext = create.createFullContext;
exports.createIndexerContext = create.createIndexerContext;
exports.createRpcContext = create.createRpcContext;
exports.ZTG = ztg.ZTG;
exports.IOAssetId = assetid.IOAssetId;
exports.IOCategoricalAssetId = assetid.IOCategoricalAssetId;
exports.IOCategoricalIndex = assetid.IOCategoricalIndex;
exports.IOMarketOutcomeAssetId = assetid.IOMarketOutcomeAssetId;
exports.IOPoolShareAssetId = assetid.IOPoolShareAssetId;
exports.IOScalarAssetId = assetid.IOScalarAssetId;
exports.IOScalarIndex = assetid.IOScalarIndex;
exports.IOScalarPosition = assetid.IOScalarPosition;
exports.IOZtgAssetId = assetid.IOZtgAssetId;
exports.fromPrimitive = assetid.fromPrimitive;
exports.getIndexOf = assetid.getIndexOf;
exports.getMarketIdOf = assetid.getMarketIdOf;
exports.getScalarIndexOf = assetid.getScalarIndexOf;
exports.parseAssetId = assetid.parseAssetId;
exports.IOBlockNumber = blocknumber.IOBlockNumber;
exports.isBlockNumber = blocknumber.isBlockNumber;
exports.IOMarketId = marketid.IOMarketId;
exports.isMarketId = marketid.isMarketId;
exports.isIndexedData = data.isIndexedData;
exports.isRpcData = data.isRpcData;
exports.NA = na.NA;
exports.isAvailable = na.isAvailable;
exports.isNA = na.isNA;
exports.na = na.na;
exports.reasonOf = na.reasonOf;
exports.ZeitgeistIpfs = storage.ZeitgeistIpfs;
Object.defineProperty(exports, 'KnownPresets', {
	enumerable: true,
	get: function () { return types$2.KnownPresets; }
});
Object.defineProperty(exports, 'SupportedParachain', {
	enumerable: true,
	get: function () { return types$2.SupportedParachain; }
});
exports.isKnownPreset = types$2.isKnownPreset;
exports.isFullContext = types$3.isFullContext;
exports.isIndexerContext = types$3.isIndexerContext;
exports.isRpcContext = types$3.isRpcContext;
exports.teardown = types$3.teardown;
exports.isWithPool = types$4.isWithPool;
exports.attachMarketMethods = market.attachMarketMethods;
exports.getDeadlines = market.getDeadlines;
exports.getPoolId = market.getPoolId;
exports.getPrediction = market.getPrediction;
exports.getReportedAt = market.getReportedAt;
exports.getReporter = market.getReporter;
exports.getScalarBounds = market.getScalarBounds;
exports.getStatus = market.getStatus;
exports.hasMarketMethods = market.hasMarketMethods;
exports.hasPool = market.hasPool;
exports.hasReport = market.hasReport;
exports.projectEndTimestamp = market.projectEndTimestamp;
exports.rpcMarket = market.rpcMarket;
exports.timespanOf = market.timespanOf;
exports.isInfinite = marketstage.isInfinite;
exports.isMarketIdQuery = types$5.isMarketIdQuery;
exports.isPoolIdQuery = types$5.isPoolIdQuery;
exports.fromEntries = pool.fromEntries;
exports.getAssetWeight = pool.getAssetWeight;
exports.rpcPool = pool.rpcPool;
