// @zeitgeistpm/sdk/dist/model/swaps/index.esm.js
import { pfunc } from '@zeitgeistpm/utility/dist/pfunc';
import { isRpcContext } from '../../context/types.esm.js';
import { saturatedPoolsIndex } from './functions/saturatedPoolsIndex/index.esm.js';
import { getPool, observePool$ } from './functions/getPool/index.esm.js';
import { listPools } from './functions/listPools/index.esm.js';
import { poolPrices, observePoolPrices$ } from './functions/poolPrices/index.esm.js';
import '@polkadot/util';
import '@zeitgeistpm/rpc';
import '@zeitgeistpm/utility/dist/option';
import '@zeitgeistpm/utility/dist/taskeither';
import 'decimal.js';
import 'lodash-es';
import '../../primitives/ztg.esm.js';
import '../../primitives/assetid.esm.js';
import '../../primitives/blocknumber.esm.js';
import '../../primitives/marketid.esm.js';
import '@zeitgeistpm/utility/dist/null';

/**
 * Create top level enriched zeitgeist Swaps model.
 *
 * @generic C - Context
 * @param ctx C
 * @returns Swaps<C>
 */
const model = (ctx) => {
    const swaps = {
        listPools: query => listPools(ctx, query),
        getPool: pfunc((query) => getPool(ctx, query), (isRpcContext(ctx)
            ? {
                $: query => observePool$(ctx, query),
            }
            : {})),
        poolPrices: pfunc((query) => poolPrices(ctx, query), (isRpcContext(ctx)
            ? {
                $: query => observePoolPrices$(ctx, query),
            }
            : {})),
        saturatedPoolsIndex: pools => saturatedPoolsIndex(ctx, pools),
    };
    return swaps;
};

export { model };
