// @zeitgeistpm/indexer/dist/graphql/sdk.js
'use strict';

var gql = require('graphql-tag');

exports.AccountBalanceOrderByInput = void 0;
(function (AccountBalanceOrderByInput) {
    AccountBalanceOrderByInput["AccountAccountIdAsc"] = "account_accountId_ASC";
    AccountBalanceOrderByInput["AccountAccountIdDesc"] = "account_accountId_DESC";
    AccountBalanceOrderByInput["AccountIdAsc"] = "account_id_ASC";
    AccountBalanceOrderByInput["AccountIdDesc"] = "account_id_DESC";
    AccountBalanceOrderByInput["AccountMarketIdAsc"] = "account_marketId_ASC";
    AccountBalanceOrderByInput["AccountMarketIdDesc"] = "account_marketId_DESC";
    AccountBalanceOrderByInput["AccountPoolIdAsc"] = "account_poolId_ASC";
    AccountBalanceOrderByInput["AccountPoolIdDesc"] = "account_poolId_DESC";
    AccountBalanceOrderByInput["AssetIdAsc"] = "assetId_ASC";
    AccountBalanceOrderByInput["AssetIdDesc"] = "assetId_DESC";
    AccountBalanceOrderByInput["BalanceAsc"] = "balance_ASC";
    AccountBalanceOrderByInput["BalanceDesc"] = "balance_DESC";
    AccountBalanceOrderByInput["IdAsc"] = "id_ASC";
    AccountBalanceOrderByInput["IdDesc"] = "id_DESC";
})(exports.AccountBalanceOrderByInput || (exports.AccountBalanceOrderByInput = {}));
exports.AccountOrderByInput = void 0;
(function (AccountOrderByInput) {
    AccountOrderByInput["AccountIdAsc"] = "accountId_ASC";
    AccountOrderByInput["AccountIdDesc"] = "accountId_DESC";
    AccountOrderByInput["IdAsc"] = "id_ASC";
    AccountOrderByInput["IdDesc"] = "id_DESC";
    AccountOrderByInput["MarketIdAsc"] = "marketId_ASC";
    AccountOrderByInput["MarketIdDesc"] = "marketId_DESC";
    AccountOrderByInput["PoolIdAsc"] = "poolId_ASC";
    AccountOrderByInput["PoolIdDesc"] = "poolId_DESC";
})(exports.AccountOrderByInput || (exports.AccountOrderByInput = {}));
exports.AssetOrderByInput = void 0;
(function (AssetOrderByInput) {
    AssetOrderByInput["AmountInPoolAsc"] = "amountInPool_ASC";
    AssetOrderByInput["AmountInPoolDesc"] = "amountInPool_DESC";
    AssetOrderByInput["AssetIdAsc"] = "assetId_ASC";
    AssetOrderByInput["AssetIdDesc"] = "assetId_DESC";
    AssetOrderByInput["IdAsc"] = "id_ASC";
    AssetOrderByInput["IdDesc"] = "id_DESC";
    AssetOrderByInput["PoolIdAsc"] = "poolId_ASC";
    AssetOrderByInput["PoolIdDesc"] = "poolId_DESC";
    AssetOrderByInput["PriceAsc"] = "price_ASC";
    AssetOrderByInput["PriceDesc"] = "price_DESC";
})(exports.AssetOrderByInput || (exports.AssetOrderByInput = {}));
exports.HistoricalAccountBalanceOrderByInput = void 0;
(function (HistoricalAccountBalanceOrderByInput) {
    HistoricalAccountBalanceOrderByInput["AccountIdAsc"] = "accountId_ASC";
    HistoricalAccountBalanceOrderByInput["AccountIdDesc"] = "accountId_DESC";
    HistoricalAccountBalanceOrderByInput["AssetIdAsc"] = "assetId_ASC";
    HistoricalAccountBalanceOrderByInput["AssetIdDesc"] = "assetId_DESC";
    HistoricalAccountBalanceOrderByInput["BalanceAsc"] = "balance_ASC";
    HistoricalAccountBalanceOrderByInput["BalanceDesc"] = "balance_DESC";
    HistoricalAccountBalanceOrderByInput["BlockNumberAsc"] = "blockNumber_ASC";
    HistoricalAccountBalanceOrderByInput["BlockNumberDesc"] = "blockNumber_DESC";
    HistoricalAccountBalanceOrderByInput["DBalanceAsc"] = "dBalance_ASC";
    HistoricalAccountBalanceOrderByInput["DBalanceDesc"] = "dBalance_DESC";
    HistoricalAccountBalanceOrderByInput["EventAsc"] = "event_ASC";
    HistoricalAccountBalanceOrderByInput["EventDesc"] = "event_DESC";
    HistoricalAccountBalanceOrderByInput["IdAsc"] = "id_ASC";
    HistoricalAccountBalanceOrderByInput["IdDesc"] = "id_DESC";
    HistoricalAccountBalanceOrderByInput["TimestampAsc"] = "timestamp_ASC";
    HistoricalAccountBalanceOrderByInput["TimestampDesc"] = "timestamp_DESC";
})(exports.HistoricalAccountBalanceOrderByInput || (exports.HistoricalAccountBalanceOrderByInput = {}));
exports.HistoricalAssetOrderByInput = void 0;
(function (HistoricalAssetOrderByInput) {
    HistoricalAssetOrderByInput["AccountIdAsc"] = "accountId_ASC";
    HistoricalAssetOrderByInput["AccountIdDesc"] = "accountId_DESC";
    HistoricalAssetOrderByInput["AssetIdAsc"] = "assetId_ASC";
    HistoricalAssetOrderByInput["AssetIdDesc"] = "assetId_DESC";
    HistoricalAssetOrderByInput["BlockNumberAsc"] = "blockNumber_ASC";
    HistoricalAssetOrderByInput["BlockNumberDesc"] = "blockNumber_DESC";
    HistoricalAssetOrderByInput["DAmountInPoolAsc"] = "dAmountInPool_ASC";
    HistoricalAssetOrderByInput["DAmountInPoolDesc"] = "dAmountInPool_DESC";
    HistoricalAssetOrderByInput["DPriceAsc"] = "dPrice_ASC";
    HistoricalAssetOrderByInput["DPriceDesc"] = "dPrice_DESC";
    HistoricalAssetOrderByInput["EventAsc"] = "event_ASC";
    HistoricalAssetOrderByInput["EventDesc"] = "event_DESC";
    HistoricalAssetOrderByInput["IdAsc"] = "id_ASC";
    HistoricalAssetOrderByInput["IdDesc"] = "id_DESC";
    HistoricalAssetOrderByInput["NewAmountInPoolAsc"] = "newAmountInPool_ASC";
    HistoricalAssetOrderByInput["NewAmountInPoolDesc"] = "newAmountInPool_DESC";
    HistoricalAssetOrderByInput["NewPriceAsc"] = "newPrice_ASC";
    HistoricalAssetOrderByInput["NewPriceDesc"] = "newPrice_DESC";
    HistoricalAssetOrderByInput["TimestampAsc"] = "timestamp_ASC";
    HistoricalAssetOrderByInput["TimestampDesc"] = "timestamp_DESC";
    HistoricalAssetOrderByInput["ZtgTradedAsc"] = "ztgTraded_ASC";
    HistoricalAssetOrderByInput["ZtgTradedDesc"] = "ztgTraded_DESC";
})(exports.HistoricalAssetOrderByInput || (exports.HistoricalAssetOrderByInput = {}));
exports.HistoricalMarketOrderByInput = void 0;
(function (HistoricalMarketOrderByInput) {
    HistoricalMarketOrderByInput["BlockNumberAsc"] = "blockNumber_ASC";
    HistoricalMarketOrderByInput["BlockNumberDesc"] = "blockNumber_DESC";
    HistoricalMarketOrderByInput["EventAsc"] = "event_ASC";
    HistoricalMarketOrderByInput["EventDesc"] = "event_DESC";
    HistoricalMarketOrderByInput["IdAsc"] = "id_ASC";
    HistoricalMarketOrderByInput["IdDesc"] = "id_DESC";
    HistoricalMarketOrderByInput["MarketIdAsc"] = "marketId_ASC";
    HistoricalMarketOrderByInput["MarketIdDesc"] = "marketId_DESC";
    HistoricalMarketOrderByInput["PoolIdAsc"] = "poolId_ASC";
    HistoricalMarketOrderByInput["PoolIdDesc"] = "poolId_DESC";
    HistoricalMarketOrderByInput["ReportAtAsc"] = "report_at_ASC";
    HistoricalMarketOrderByInput["ReportAtDesc"] = "report_at_DESC";
    HistoricalMarketOrderByInput["ReportByAsc"] = "report_by_ASC";
    HistoricalMarketOrderByInput["ReportByDesc"] = "report_by_DESC";
    HistoricalMarketOrderByInput["ResolvedOutcomeAsc"] = "resolvedOutcome_ASC";
    HistoricalMarketOrderByInput["ResolvedOutcomeDesc"] = "resolvedOutcome_DESC";
    HistoricalMarketOrderByInput["StatusAsc"] = "status_ASC";
    HistoricalMarketOrderByInput["StatusDesc"] = "status_DESC";
    HistoricalMarketOrderByInput["TimestampAsc"] = "timestamp_ASC";
    HistoricalMarketOrderByInput["TimestampDesc"] = "timestamp_DESC";
})(exports.HistoricalMarketOrderByInput || (exports.HistoricalMarketOrderByInput = {}));
exports.HistoricalPoolOrderByInput = void 0;
(function (HistoricalPoolOrderByInput) {
    HistoricalPoolOrderByInput["BlockNumberAsc"] = "blockNumber_ASC";
    HistoricalPoolOrderByInput["BlockNumberDesc"] = "blockNumber_DESC";
    HistoricalPoolOrderByInput["DVolumeAsc"] = "dVolume_ASC";
    HistoricalPoolOrderByInput["DVolumeDesc"] = "dVolume_DESC";
    HistoricalPoolOrderByInput["EventAsc"] = "event_ASC";
    HistoricalPoolOrderByInput["EventDesc"] = "event_DESC";
    HistoricalPoolOrderByInput["IdAsc"] = "id_ASC";
    HistoricalPoolOrderByInput["IdDesc"] = "id_DESC";
    HistoricalPoolOrderByInput["PoolIdAsc"] = "poolId_ASC";
    HistoricalPoolOrderByInput["PoolIdDesc"] = "poolId_DESC";
    HistoricalPoolOrderByInput["PoolStatusAsc"] = "poolStatus_ASC";
    HistoricalPoolOrderByInput["PoolStatusDesc"] = "poolStatus_DESC";
    HistoricalPoolOrderByInput["TimestampAsc"] = "timestamp_ASC";
    HistoricalPoolOrderByInput["TimestampDesc"] = "timestamp_DESC";
    HistoricalPoolOrderByInput["VolumeAsc"] = "volume_ASC";
    HistoricalPoolOrderByInput["VolumeDesc"] = "volume_DESC";
    HistoricalPoolOrderByInput["ZtgQtyAsc"] = "ztgQty_ASC";
    HistoricalPoolOrderByInput["ZtgQtyDesc"] = "ztgQty_DESC";
})(exports.HistoricalPoolOrderByInput || (exports.HistoricalPoolOrderByInput = {}));
exports.MarketOrderByInput = void 0;
(function (MarketOrderByInput) {
    MarketOrderByInput["AuthorizedAddressAsc"] = "authorizedAddress_ASC";
    MarketOrderByInput["AuthorizedAddressDesc"] = "authorizedAddress_DESC";
    MarketOrderByInput["BaseAssetAsc"] = "baseAsset_ASC";
    MarketOrderByInput["BaseAssetDesc"] = "baseAsset_DESC";
    MarketOrderByInput["CreationAsc"] = "creation_ASC";
    MarketOrderByInput["CreationDesc"] = "creation_DESC";
    MarketOrderByInput["CreatorFeeAsc"] = "creatorFee_ASC";
    MarketOrderByInput["CreatorFeeDesc"] = "creatorFee_DESC";
    MarketOrderByInput["CreatorAsc"] = "creator_ASC";
    MarketOrderByInput["CreatorDesc"] = "creator_DESC";
    MarketOrderByInput["DeadlinesDisputeDurationAsc"] = "deadlines_disputeDuration_ASC";
    MarketOrderByInput["DeadlinesDisputeDurationDesc"] = "deadlines_disputeDuration_DESC";
    MarketOrderByInput["DeadlinesGracePeriodAsc"] = "deadlines_gracePeriod_ASC";
    MarketOrderByInput["DeadlinesGracePeriodDesc"] = "deadlines_gracePeriod_DESC";
    MarketOrderByInput["DeadlinesOracleDurationAsc"] = "deadlines_oracleDuration_ASC";
    MarketOrderByInput["DeadlinesOracleDurationDesc"] = "deadlines_oracleDuration_DESC";
    MarketOrderByInput["DescriptionAsc"] = "description_ASC";
    MarketOrderByInput["DescriptionDesc"] = "description_DESC";
    MarketOrderByInput["DisputeMechanismAsc"] = "disputeMechanism_ASC";
    MarketOrderByInput["DisputeMechanismDesc"] = "disputeMechanism_DESC";
    MarketOrderByInput["IdAsc"] = "id_ASC";
    MarketOrderByInput["IdDesc"] = "id_DESC";
    MarketOrderByInput["ImgAsc"] = "img_ASC";
    MarketOrderByInput["ImgDesc"] = "img_DESC";
    MarketOrderByInput["MarketIdAsc"] = "marketId_ASC";
    MarketOrderByInput["MarketIdDesc"] = "marketId_DESC";
    MarketOrderByInput["MarketTypeCategoricalAsc"] = "marketType_categorical_ASC";
    MarketOrderByInput["MarketTypeCategoricalDesc"] = "marketType_categorical_DESC";
    MarketOrderByInput["MetadataAsc"] = "metadata_ASC";
    MarketOrderByInput["MetadataDesc"] = "metadata_DESC";
    MarketOrderByInput["OracleAsc"] = "oracle_ASC";
    MarketOrderByInput["OracleDesc"] = "oracle_DESC";
    MarketOrderByInput["PeriodEndAsc"] = "period_end_ASC";
    MarketOrderByInput["PeriodEndDesc"] = "period_end_DESC";
    MarketOrderByInput["PeriodStartAsc"] = "period_start_ASC";
    MarketOrderByInput["PeriodStartDesc"] = "period_start_DESC";
    MarketOrderByInput["PoolAccountIdAsc"] = "pool_accountId_ASC";
    MarketOrderByInput["PoolAccountIdDesc"] = "pool_accountId_DESC";
    MarketOrderByInput["PoolBaseAssetAsc"] = "pool_baseAsset_ASC";
    MarketOrderByInput["PoolBaseAssetDesc"] = "pool_baseAsset_DESC";
    MarketOrderByInput["PoolCreatedAtAsc"] = "pool_createdAt_ASC";
    MarketOrderByInput["PoolCreatedAtDesc"] = "pool_createdAt_DESC";
    MarketOrderByInput["PoolIdAsc"] = "pool_id_ASC";
    MarketOrderByInput["PoolIdDesc"] = "pool_id_DESC";
    MarketOrderByInput["PoolMarketIdAsc"] = "pool_marketId_ASC";
    MarketOrderByInput["PoolMarketIdDesc"] = "pool_marketId_DESC";
    MarketOrderByInput["PoolPoolIdAsc"] = "pool_poolId_ASC";
    MarketOrderByInput["PoolPoolIdDesc"] = "pool_poolId_DESC";
    MarketOrderByInput["PoolPoolStatusAsc"] = "pool_poolStatus_ASC";
    MarketOrderByInput["PoolPoolStatusDesc"] = "pool_poolStatus_DESC";
    MarketOrderByInput["PoolScoringRuleAsc"] = "pool_scoringRule_ASC";
    MarketOrderByInput["PoolScoringRuleDesc"] = "pool_scoringRule_DESC";
    MarketOrderByInput["PoolSwapFeeAsc"] = "pool_swapFee_ASC";
    MarketOrderByInput["PoolSwapFeeDesc"] = "pool_swapFee_DESC";
    MarketOrderByInput["PoolTotalSubsidyAsc"] = "pool_totalSubsidy_ASC";
    MarketOrderByInput["PoolTotalSubsidyDesc"] = "pool_totalSubsidy_DESC";
    MarketOrderByInput["PoolTotalWeightAsc"] = "pool_totalWeight_ASC";
    MarketOrderByInput["PoolTotalWeightDesc"] = "pool_totalWeight_DESC";
    MarketOrderByInput["PoolVolumeAsc"] = "pool_volume_ASC";
    MarketOrderByInput["PoolVolumeDesc"] = "pool_volume_DESC";
    MarketOrderByInput["PoolZtgQtyAsc"] = "pool_ztgQty_ASC";
    MarketOrderByInput["PoolZtgQtyDesc"] = "pool_ztgQty_DESC";
    MarketOrderByInput["QuestionAsc"] = "question_ASC";
    MarketOrderByInput["QuestionDesc"] = "question_DESC";
    MarketOrderByInput["RejectReasonAsc"] = "rejectReason_ASC";
    MarketOrderByInput["RejectReasonDesc"] = "rejectReason_DESC";
    MarketOrderByInput["ReportAtAsc"] = "report_at_ASC";
    MarketOrderByInput["ReportAtDesc"] = "report_at_DESC";
    MarketOrderByInput["ReportByAsc"] = "report_by_ASC";
    MarketOrderByInput["ReportByDesc"] = "report_by_DESC";
    MarketOrderByInput["ResolvedOutcomeAsc"] = "resolvedOutcome_ASC";
    MarketOrderByInput["ResolvedOutcomeDesc"] = "resolvedOutcome_DESC";
    MarketOrderByInput["ScalarTypeAsc"] = "scalarType_ASC";
    MarketOrderByInput["ScalarTypeDesc"] = "scalarType_DESC";
    MarketOrderByInput["ScoringRuleAsc"] = "scoringRule_ASC";
    MarketOrderByInput["ScoringRuleDesc"] = "scoringRule_DESC";
    MarketOrderByInput["SlugAsc"] = "slug_ASC";
    MarketOrderByInput["SlugDesc"] = "slug_DESC";
    MarketOrderByInput["StatusAsc"] = "status_ASC";
    MarketOrderByInput["StatusDesc"] = "status_DESC";
})(exports.MarketOrderByInput || (exports.MarketOrderByInput = {}));
exports.PoolOrderByInput = void 0;
(function (PoolOrderByInput) {
    PoolOrderByInput["AccountIdAsc"] = "accountId_ASC";
    PoolOrderByInput["AccountIdDesc"] = "accountId_DESC";
    PoolOrderByInput["BaseAssetAsc"] = "baseAsset_ASC";
    PoolOrderByInput["BaseAssetDesc"] = "baseAsset_DESC";
    PoolOrderByInput["CreatedAtAsc"] = "createdAt_ASC";
    PoolOrderByInput["CreatedAtDesc"] = "createdAt_DESC";
    PoolOrderByInput["IdAsc"] = "id_ASC";
    PoolOrderByInput["IdDesc"] = "id_DESC";
    PoolOrderByInput["MarketIdAsc"] = "marketId_ASC";
    PoolOrderByInput["MarketIdDesc"] = "marketId_DESC";
    PoolOrderByInput["PoolIdAsc"] = "poolId_ASC";
    PoolOrderByInput["PoolIdDesc"] = "poolId_DESC";
    PoolOrderByInput["PoolStatusAsc"] = "poolStatus_ASC";
    PoolOrderByInput["PoolStatusDesc"] = "poolStatus_DESC";
    PoolOrderByInput["ScoringRuleAsc"] = "scoringRule_ASC";
    PoolOrderByInput["ScoringRuleDesc"] = "scoringRule_DESC";
    PoolOrderByInput["SwapFeeAsc"] = "swapFee_ASC";
    PoolOrderByInput["SwapFeeDesc"] = "swapFee_DESC";
    PoolOrderByInput["TotalSubsidyAsc"] = "totalSubsidy_ASC";
    PoolOrderByInput["TotalSubsidyDesc"] = "totalSubsidy_DESC";
    PoolOrderByInput["TotalWeightAsc"] = "totalWeight_ASC";
    PoolOrderByInput["TotalWeightDesc"] = "totalWeight_DESC";
    PoolOrderByInput["VolumeAsc"] = "volume_ASC";
    PoolOrderByInput["VolumeDesc"] = "volume_DESC";
    PoolOrderByInput["ZtgQtyAsc"] = "ztgQty_ASC";
    PoolOrderByInput["ZtgQtyDesc"] = "ztgQty_DESC";
})(exports.PoolOrderByInput || (exports.PoolOrderByInput = {}));
const FullAccountBalanceFragmentDoc = gql `
    fragment FullAccountBalance on AccountBalance {
  account {
    accountId
    id
    marketId
    poolId
  }
  assetId
  balance
  id
}
    `;
const FullAssetFragmentDoc = gql `
    fragment FullAsset on Asset {
  id
  assetId
  poolId
  price
  amountInPool
}
    `;
const FullHistoricalAccountBalanceFragmentDoc = gql `
    fragment FullHistoricalAccountBalance on HistoricalAccountBalance {
  accountId
  assetId
  balance
  blockNumber
  dBalance
  event
  id
  timestamp
}
    `;
const FullHistoricalAssetsFragmentDoc = gql `
    fragment FullHistoricalAssets on HistoricalAsset {
  accountId
  assetId
  blockNumber
  dAmountInPool
  dPrice
  event
  id
  newAmountInPool
  newPrice
  timestamp
  ztgTraded
}
    `;
const FullPoolFragmentDoc = gql `
    fragment FullPool on Pool {
  accountId
  baseAsset
  createdAt
  id
  marketId
  poolId
  poolStatus
  scoringRule
  swapFee
  totalSubsidy
  totalWeight
  volume
  weights {
    assetId
    len
  }
  ztgQty
}
    `;
const FullMarketFragmentDoc = gql `
    fragment FullMarket on Market {
  id
  marketId
  description
  creator
  creatorFee
  creation
  oracle
  question
  slug
  img
  tags
  status
  scoringRule
  resolvedOutcome
  scalarType
  outcomeAssets
  rejectReason
  marketType {
    categorical
    scalar
  }
  period {
    block
    timestamp
    end
    start
  }
  report {
    at
    by
    outcome {
      categorical
      scalar
    }
  }
  disputeMechanism
  categories {
    ticker
    name
    color
  }
  deadlines {
    disputeDuration
    gracePeriod
    oracleDuration
  }
  bonds {
    creation {
      isSettled
      value
      who
    }
    oracle {
      isSettled
      value
      who
    }
  }
  pool {
    ...FullPool
  }
}
    ${FullPoolFragmentDoc}`;
const AccountBalancesDocument = gql `
    query accountBalances($where: AccountBalanceWhereInput, $order: [AccountBalanceOrderByInput!], $offset: Int, $limit: Int) {
  accountBalances(where: $where, orderBy: $order, offset: $offset, limit: $limit) {
    ...FullAccountBalance
  }
}
    ${FullAccountBalanceFragmentDoc}`;
const AssetsDocument = gql `
    query assets($where: AssetWhereInput, $order: [AssetOrderByInput!], $offset: Int, $limit: Int) {
  assets(where: $where, orderBy: $order, offset: $offset, limit: $limit) {
    ...FullAsset
  }
}
    ${FullAssetFragmentDoc}`;
const HistoricalAccountBalancesDocument = gql `
    query historicalAccountBalances($where: HistoricalAccountBalanceWhereInput, $order: [HistoricalAccountBalanceOrderByInput!], $offset: Int, $limit: Int) {
  historicalAccountBalances(
    where: $where
    orderBy: $order
    offset: $offset
    limit: $limit
  ) {
    ...FullHistoricalAccountBalance
  }
}
    ${FullHistoricalAccountBalanceFragmentDoc}`;
const HistoricalAssetsDocument = gql `
    query historicalAssets($where: HistoricalAssetWhereInput, $order: [HistoricalAssetOrderByInput!], $offset: Int, $limit: Int) {
  historicalAssets(where: $where, orderBy: $order, offset: $offset, limit: $limit) {
    ...FullHistoricalAssets
  }
}
    ${FullHistoricalAssetsFragmentDoc}`;
const MarketsDocument = gql `
    query markets($where: MarketWhereInput, $order: [MarketOrderByInput!], $limit: Int, $offset: Int) {
  markets(where: $where, limit: $limit, offset: $offset, orderBy: $order) {
    ...FullMarket
  }
}
    ${FullMarketFragmentDoc}`;
const MarketStatusCountDocument = gql `
    query marketStatusCount($status: String!) {
  markets(where: {status_eq: $status}) {
    id
  }
}
    `;
const PingQueryDocument = gql `
    query pingQuery {
  markets(limit: 1) {
    id
  }
}
    `;
const PoolsDocument = gql `
    query pools($where: PoolWhereInput, $order: [PoolOrderByInput!], $offset: Int, $limit: Int) {
  pools(where: $where, orderBy: $order, offset: $offset, limit: $limit) {
    ...FullPool
  }
}
    ${FullPoolFragmentDoc}`;
const defaultWrapper = (action, _operationName, _operationType) => action();
function getSdk(client, withWrapper = defaultWrapper) {
    return {
        accountBalances(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(AccountBalancesDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'accountBalances', 'query');
        },
        assets(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(AssetsDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'assets', 'query');
        },
        historicalAccountBalances(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(HistoricalAccountBalancesDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'historicalAccountBalances', 'query');
        },
        historicalAssets(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(HistoricalAssetsDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'historicalAssets', 'query');
        },
        markets(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(MarketsDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'markets', 'query');
        },
        marketStatusCount(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(MarketStatusCountDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'marketStatusCount', 'query');
        },
        pingQuery(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(PingQueryDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'pingQuery', 'query');
        },
        pools(variables, requestHeaders) {
            return withWrapper((wrappedRequestHeaders) => client.request(PoolsDocument, variables, { ...requestHeaders, ...wrappedRequestHeaders }), 'pools', 'query');
        }
    };
}

exports.AccountBalancesDocument = AccountBalancesDocument;
exports.AssetsDocument = AssetsDocument;
exports.FullAccountBalanceFragmentDoc = FullAccountBalanceFragmentDoc;
exports.FullAssetFragmentDoc = FullAssetFragmentDoc;
exports.FullHistoricalAccountBalanceFragmentDoc = FullHistoricalAccountBalanceFragmentDoc;
exports.FullHistoricalAssetsFragmentDoc = FullHistoricalAssetsFragmentDoc;
exports.FullMarketFragmentDoc = FullMarketFragmentDoc;
exports.FullPoolFragmentDoc = FullPoolFragmentDoc;
exports.HistoricalAccountBalancesDocument = HistoricalAccountBalancesDocument;
exports.HistoricalAssetsDocument = HistoricalAssetsDocument;
exports.MarketStatusCountDocument = MarketStatusCountDocument;
exports.MarketsDocument = MarketsDocument;
exports.PingQueryDocument = PingQueryDocument;
exports.PoolsDocument = PoolsDocument;
exports.getSdk = getSdk;
