// @zeitgeistpm/rpc/dist/index.js
'use strict';

var augmentApi = require('@zeitgeistpm/augment-api');
var types = require('./lib/transactions/types.js');
var index = require('./lib/keyring/index.js');
var index$1 = require('./lib/transactions/index.js');

/**
 * Create options including types and typedefs for the zeitgeist rpc api.
 *
 * @param opts ZeitgeistApiOptions
 * @returns ApiOptions
 */
const options = (opts) => {
    return {
        types: {
            ...Object.values(augmentApi.definitions).reduce((acc, { types }) => ({ ...acc, ...types }), {}),
        },
        rpc: {
            ...Object.entries(augmentApi.definitions).reduce((acc, [key, { rpc }]) => ({ ...acc, [key]: rpc }), {}),
        },
        provider: opts.provider,
        noInitWarn: true,
    };
};

exports.RetractedError = types.RetractedError;
exports.UnknownDispatchError = types.UnknownDispatchError;
exports.transactionErrorToString = types.transactionErrorToString;
exports.isExtSigner = index.isExtSigner;
exports.signAndSend = index$1.signAndSend;
exports.options = options;
